packages/kbn-esql-validation-autocomplete/src/autocomplete/commands/sort/helper.ts

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+const regexStart = /.+\|\s*so?r?(?<start>t?)(.+,)?(?<space1>\s+)?/i;
+const regex =
+  /.+\|\s*sort(.+,)?((?<space1>\s+)(?<column>[^\s]+)(?<space2>\s*)(?<order>(AS?C?)|(DE?S?C?))?(?<space3>\s*)(?<nulls>NU?L?L?S? ?(FI?R?S?T?|LA?S?T?)?)?(?<space4>\s*))?/i;
```

Comment:
I like this experiment with giving `sort` its own autocomplete routine. But, I'm wondering if resorting to a regex is a bit of a regression because haven't we already done the parsing to build the AST?

Reply:
I don't think it is a regression, it is an improvement, IMO. Our current query text auto-correct + parser does not handle these cases.

For example, if you have a text like:

```
from a | sort stringField A
```

Our query text auto-correct will simply insert a marker, like so:

```
from a | sort stringField Amarker_esql_editor
```

which will result in an invalid query, and there is no information about the fact that user has already typed `A`, as for `ASC`.

Also, because the way the marker is inserted, it results into an word `Amarker_esql_editor`, which&mdash;if the query was actually parsed&mdash;would fail to remove that marker from the query afterwards.

---

Basically, our current autocomplete does not handle this situation well at all. The regex is used *only to detect the position* inside the `SORT` command, once we know the position, we can provide autocomplete suggestions.

Reply:
I agree that this is an improvement in functionality. By regression, I meant a step backward technically. In other words, I'm not sure this is the direction we should take.

You've correctly identified the `marker_esql_editor` as a major problem. This is because it is being added in instances where it shouldn't be and then AST nodes are being incorrectly removed. If this is fixed, I am expecting that we will be in a much better position to leverage the AST because we will have the AST nodes available in many more cases.

The direction I would like to see us move is away from string scanning toward leveraging the AST where possible.

I don't necessarily mind this string scanning as a short-term solution, but if we aren't aligned on the overall direction away from it, we should talk more.

Reply:
> You've correctly identified the `marker_esql_editor` as a major problem. This is because it is being added in instances where it shouldn't be and then AST nodes are being incorrectly removed. If this is fixed, I am expecting that we will be in a much better position to leverage the AST because we will have the AST nodes available in many more cases.

I agree that we should use the parsed AST where possible. But our autocompletion AST: (1) is not parsed correctly in all cases (see above note on `Amarker_esql_editor`); even if it was parsed (2) we still need to detect the position *within* the order node.

I'm not sure how the current approach with the marker insertion and parsing can solve the above. Even if AST was parsed somehow correctly, there is not enough information in the AST to find the location within the node:

```
   * SORT <column> [ASC/DESC] [NULLS FIRST/NULLS LAST]
   * |  | |       |   |      |      |                 |
   * |  | |       |   |      |      |                 space4
   * |  | |       |   |      |      nulls
   * |  | |       |   |      space3
   * |  | |       |   order
   * |  | |       space 2
   * |  | |
   * |  | column
   * |  start
   * pre-start
```

At least that is my understanding, wdyt?

Reply:
The ANTLR parsing output, the CST should have all the needed information, we could use that finding the within-node position, but I'm thinking that would be messier, especially because ES team can change the token names of the ANTLR grammar at will.

Reply:
> This PR is using the AST to find the node in question (the SORT command in this case). But, once the node is found, I want to find the position within the node. The AST does not have much within-node information, so the regular expression is used to find the position within the node.

This makes sense. Thank you for discussing.

> The ANTLR parsing output, the CST, should have all the needed information, we could use that for finding the within-node position, but I'm thinking that would be messier, especially because ES team can change the token names of the ANTLR grammar at will.

Agreed. We want the abstraction.